src/vsock/src/rxqueue.rs

```diff
@@ -0,0 +1,156 @@
+#![deny(missing_docs)]
+use super::RxOps;
+
+#[derive(Debug)]
+pub struct RxQueue {
+    /// bitmap of rx operations
+    queue: u8,
+}
+
+impl RxQueue {
+    /// new instance of RxQueue
+    pub fn new() -> Self {
+        RxQueue { queue: 0_u8 }
+    }
+
+    /// Enqueue a new rx operation into the queue
+    pub fn enqueue(&mut self, op: RxOps) {
+        self.queue |= op.bitmask();
+    }
+
+    /// Dequeue an rx operation from the queue
+    pub fn dequeue(&mut self) -> Option<RxOps> {
+        match self.peek() {
+            Some(req) => {
+                self.queue &= !req.bitmask();
+                Some(req)
+            }
+            None => None,
+        }
+    }
+
+    /// Peek into the queue to check if it contains an rx operation
+    pub fn peek(&self) -> Option<RxOps> {
+        if self.contains(RxOps::Request.bitmask()) {
+            return Some(RxOps::Request);
+        }
+        if self.contains(RxOps::Rw.bitmask()) {
+            return Some(RxOps::Rw);
+        }
+        if self.contains(RxOps::Response.bitmask()) {
+            return Some(RxOps::Response);
+        }
+        if self.contains(RxOps::CreditUpdate.bitmask()) {
+            return Some(RxOps::CreditUpdate);
+        }
+        if self.contains(RxOps::Reset.bitmask()) {
+            Some(RxOps::Reset)
+        } else {
+            None
+        }
```

Comment:
Can this be implemented with match instead ?

Reply:
@vireshk I am not sure about that here as I reuse the "contains" method to perform the peek. Do you have any idea how it can be achieved using a match?

Reply:
Before that I must say I am getting a bit more confused here :)

If I look at enqueue() and dequeue(), it looks like self.queue can hold multiple RxOps together. But if I look at peek(), it returns the first value set in there.

Can you clarify on how this is supposed to be used first ?

Reply:
self.queue is just a bitmap, so it stores a single operation of each kind(the location in the bitmap specifies the kind of request) for each connection. For example it can hold a single request for credit_update or OP_REQUEST etc. This is ok as each connection is "connection oriented" and processes a single request of each kind at any given time. So enqueue and dequeue push and pop operations into the queue by setting and un-setting a bit respectively. Hence, we need peek to just look at the first available operation and return that.